parlai/agents/transformer/modules/interfaces.py

```diff
@@ -0,0 +1,147 @@
+
+```python
+self.swappables.component()
+```
+
+When instantiating the model, swap out the component like so:
+
+```python
+model = SomeModel.with_components(component=NewCustomClass)()
+```
+"""
+from __future__ import annotations
+from dataclasses import dataclass
+from torch import nn
+from typing import Callable, Generic, Optional, Type, TypeVar, Union
+
+
+class ModularComponent(nn.Module):
+    """
+    A transformer component, optionally with swappable subcomponents.
+
+    For an example of how to use this, see parlai/agents/examples/transformer_variant.py
+    """
+
+    @dataclass
+    class Subcomponents:
+        """
+        Define any swappable components by adding their class as a parameter of this
+        object.
+        """
+
+    def __init__(self, *args, components: Optional[Subcomponents] = None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.swappables = components or type(self).Subcomponents()
+        assert (
+            type(self.swappables) is not ModularComponent.Subcomponents
+        ), "ModularComponent objects must define their own Subcomponents"
+
+    @classmethod
+    def with_components(cls, **kwargs) -> ModularComponentWrapper:
+        return ModularComponentWrapper(
+            klass=cls, components=cls.Subcomponents(**kwargs)
+        )
+
+
+MC = TypeVar('MC', bound=ModularComponent)
+
+
+def modular_type(klass: Type[ModularComponent]):
+    """
+    Convenience function for type annotating fields that could be either a specific
+    ModularComponent subclass or a ModularComponentWrapper around it.
+    """
+    if issubclass(klass, ModularComponent):
+        return Union[Type[klass], ModularComponentWrapper[klass]]
+    return Type[klass]
+
+
+# ModularComponentBuilder should be immutable, since it is used in default
+# function parameter values. In python, mutating a param default value changes
+# it for all subsequent invocations.
+class ModularComponentWrapper(Generic[MC]):
+    """
+    When a component has swappable subcomponents, use this object to specify both the
+    component type and it's subcomponent types at the same time.
+    """
+
+    def __init__(
+        self,
+        klass: Type[MC],
+        components: Optional[ModularComponent.Subcomponents] = None,
+    ) -> None:
+        self._cls = klass
+        self._components = components or klass.Subcomponents()
+
+    def __call__(self, *args, **kwargs) -> MC:
+        _kwargs = {'components': self._components}
+        _kwargs.update(kwargs)
+        return self._cls(*args, **_kwargs)
+
+
+def swappable(**kwargs) -> Callable[[Type], Type[MC]]:
+    """
+    Decorator for annotating a class as having swappable subcomponents.
+
+    Usage:
+
+    ```python
+    @swappable(component=DefaultClass, ...)
+    ```
+    """
+
+    def wrap(cls) -> Type:
+        return _make_class_swappable(cls, **kwargs)
+
+    return wrap
+
+
+def _make_class_swappable(cls: Type, **kwargs) -> Type[MC]:
```

Comment:
can you leave some short internal docs (comments) describing how this whole thing works?